### PR TITLE
BREAKING(std/log): simplification rewrite

### PR DESCRIPTION
**Pull request by timreichen**
#7211

- replace `setup` and global state by access to `Logger` instances to create default logger
- simplify by passing `Handler` objects instead of strings
- rename msg to message (like in [Error.prototype.message](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/message))
- add LogLevelNames enum
- rename logger.warning to warn (like in [console.warn](https://developer.mozilla.org/en-US/docs/Web/API/Console/warn)) and LogLevelNames.WARN
- remove string formatter in favor of template strings
- fix `unknown` Logger and string setup problem

### CODE REVIEW

**Comment by vwkd** [comment]
Like it, except for the reduction of error levels from six to four. More granularity is needed in production applications which only rely in logging for debugging. For a simple consumer that doesn't use all levels more granular ones are invisible and don't obstruct. Please restore the previous six levels.

File: std/log/logger.ts
Comment (vwkd) [inline std/log/logger.ts:39]: Shouldn't this be `LogLevelNames.Debug`?

Also the README should probably mention what the default log level is.

Comment (vwkd) [reply std/log/logger.ts:39]: Also, this changed from what was previously `INFO`. Why? Please revert.

**Review by vwkd — CHANGES_REQUESTED** [review]

File: std/log/logger.ts
Comment (vwkd) [inline std/log/logger.ts:42]: Please revert the default to `DEFAULT_HANDLERS = []`. Otherwise you can not create a silent logger.

One may want to do this, if one wants to log only if the program is run in certain conditions, e.g. environment flags, etc. One can not just not initialise a logger, because all the logging statements in the code would throw an error.

Comment (vwkd) [inline std/log/logger.ts:42]: Please revert the default to `DEFAULT_HANDLERS = []`. Otherwise you can not create a silent logger.

One may want to do this, if one wants to log conditionally, for example if the program is run in certain environments. One can not just not initialise a logger, because all the logging statements in the code would then throw.

**Review by vwkd — CHANGES_REQUESTED** [review]

File: std/log/logger.ts
Comment (timreichen) [reply std/log/logger.ts:42]: Hei @vwkd  thanks for your feedback. I changed the default log level to debug because imo the logger Should function similar to console by default, logging everything by default and adjust that by setting a specific log level.

Concerning the default handler: to create a silent logger, you can pass an empty array as handlers. I think this is much more clear instead of having the default logger do nothing.

Comment (vwkd) [reply std/log/logger.ts:42]: Thanks @timreichen passing an empty array to `new Logger({ handlers: [] })` actually works. Consider my previous comment as obsolete.

Regarding the default log level from my other comment: A logger framework is not intended to be similar to `console` or trying to replace it, because it has a entirely different use case. Logging for real time inspection (with `console`) and logging for retrospective inspection (with a logging library) are two very different things. It just so happens that a logging library usually also has a output channel to the console, but that is only one out of many, like for example file based logging. In no other language you'll find that the logging framework aligns to the console output, e.g. `print` in python. I'm not in favor of aligning the logging library with `console` because it has an entirely different use case.

File: std/log/logger.ts
Comment (timreichen) [reply std/log/logger.ts:39]: @vwkd actually I found that the default log level is different for different logging frameworks.
In java its `Info`, for rust and logLevel its `warn`, though they have  I just think the default should be all inclusive and if some messages should not be logged, there should be a specific log level set.



Yes, it should be a `LogLevelNames` property instead of a string.

**Comment by timreichen** [comment]
@vwkd 
Concerning `trace`: I think we could add that as a level + methods. This is also consistent with `console`.
Concerning `critical`: What would the use case be? To me a critical error would be an uncaught throw, that stops the process. If it is uncaught, it is not really loggable. If it is caught, we can use `error`. Rust seems to work fine without a `critical` level (https://docs.rs/log/0.4.1/log/enum.Level.html).

**Comment by vwkd** [comment]
@timreichen The use cases will depend on the application. I use `CRITICAL` when the application can't recover and needs to exit, while `ERROR` when the application can recover and continue. Might be worth reading: [When to use the different log levels](https://stackoverflow.com/questions/2031163/when-to-use-the-different-log-levels). But as I mentioned in my first comment, the error levels are opt-in and invisible to anyone that doesn't use them, so the more granularity the better.

File: std/log/logger.ts
Comment (timreichen) [reply std/log/logger.ts:42]: @vwkd thanks for the explanations.
### on console alignment
Maybe I did not explain well enough what I meant with `console` alignment. This module should definitely not be a replacement to `console`.
I think std modules are supposed to provide certain functionality as bear bone as possible without or the minimal possible opinion so it does not become a framework on its own.
Since there are so many different logging frameworks that not or only partially agree on log levels, namings etc. I took `console` as a guide for it is a standard web API.

### on python implementation
On research there are many frameworks that have different [log level defaults](https://github.com/denoland/deno/pull/7293#discussion_r506958120) and have  no `critical` log level at all or call it `fatal` instead of `critical`. I do not see why we should follow pythons implementation of logger other than it was the predecessor of this PR.  Or are there any specific reasons for it?
_Side note_: I would agree on implementing `trace` as a log level, but strictly speaking it is not a python log level.

### on default log level
Since different language/frameworks have different default log levels a easy and non-opinionated solution would be to not have a default one at all but request the  a specific log level in instance initialization instead.
### on default handler
I might agree on that one. Maybe it is not right to have consoleHandler as a default handler. If there was no default log level, it would be consistent not to have a default handler too.

What would you think of adjusting the `Logger` class to be simple to extend with custom log levels?

**Comment by timreichen** [comment]
@ry what is your opinion on the log levels and namings? We haven't found consensus yet on this rewrite and need some advice.

I could imagine
- `trace`
- `info`
- `debug`
- `warn`
- `error`

to be consistent with [some frameworks](https://github.com/denoland/deno/pull/7293#discussion_r507152556) and also the console API while @vwkd thinks an additional `critical` log level would be necessary. 

And if there had to be a `critical` log level: How to justify `critical` instead of `fatal`?

**Comment by jcc10** [comment]
So, I would just like to add in a completely external opinion on the critical/fatal/error/warn differentiation:

(For sake of simplicity, I'm using `request` feel free to replace with some other synonym.)

Warn: Issue handling request, request still went through. (Malformed request with extra fields that don't belong.)
Error: Issue handling request, request fails. (Malformed request missing required fields.)
Critical: Thread/Worker crash/data corruption. Isolated to one or several threads. Highly likely that other error logs will stem from this one and probably should be ignored until this is handled.
Fatal: Program crash/data corruption. The program is dead. it may not know it yet however and may dump a bunch of error logs before realizing it.

Both Critical and Fatal would probably be logging levels I would put on lines that should be impossible to reach. Places where if it does fire, I know I have some major problems on my hands.

Personally I would leave them in since it's always easier to not use stuff then to figure out how to add stuff in yourself. But I can understand the reason to remove them.

(again, just my own personal opinion on the topic, feel free to ignore :sweat_smile:  .)

**Comment by bartlomieju** [comment]
@timreichen sorry for slow review, I will try to land this PR for `1.6.0`. Please rebase the branch.

**Comment by vwkd** [comment]
I sincerely hope I'm not the only one that believes a logging framework is distinct from that of console. The line can be blurry for CLI apps, where both the user interface and the developer channel overlap, since both are the console. But you need to differentiate developer machine and client. On the client, you can not assume that that you as a developer can read the console. Logging to the console on the client makes zero sense. ESLint has even a special linting rule [no-console](https://eslint.org/docs/rules/no-console) just for this.

There should be **zero reasons** why a logging framework needs to be aligned to the console.

In concrete, this PR should **change the following**:
- Keep the six logging levels, as before. Except for renaming if desired.
- Keep no default logging handler, as before. Logging to the console on the client by default is unreasonable.

**Comment by timreichen** [comment]
@bartlomieju Thanks for the review I'll do that.
@vwkd I compared some more frameworks across multiple languages and came across the explanation, that critical/fatal would be unrecoverable and therefore end the current process (something like `Deno.exit`) immediately.
To settle the debate and solving the issue, I would propose the following:

Let 
- trace
- info
- debug
- warn
- error

be the default log levels. This is also the common ground for almost all frameworks I looked at.

Allow custom log levels to solve critical/fatal issue as well as providing more flexibility.

I created a working mockup. The syntax looks as follows:

### Default log level usage
```ts
import { logLevels, ConsoleHandler, Logger } from "https://deno.land/std/log/mod.ts"

const consoleHandler = new ConsoleHandler(logLevels.trace)
const logger = new Logger(logLevels.trace, { handlers: [consoleHandler] })
logger.trace("trace")
logger.debug("debug")
logger.info("info")
logger.warn("warn")
logger.error("error")
```
### custom `fatal` log level
```ts
import { bold, red } from "https://deno.land/std/fmt/colors.ts"
import { logLevels, ConsoleHandler, Logger } from "https://deno.land/std/log/mod.ts"

const customLogLevels = {
  ...logLevels,
  // add fatal as a LogLevel with name "Fatal" and code 60
  fatal: new LogLevel("Fatal", 60),
}

class CustomLogger extends Logger {
  // add fatal as a callable method
  fatal(message: unknown, ...args: unknown[]) {
    this.dispatch(customLogLevels.fatal, message, ...args)
    Deno.exit(1)
  }
}

const customConsoleHandler = new ConsoleHandler(customLogLevels.info)
// add fatal LogLevel method to handler
customConsoleHandler.addLogLevel(customLogLevels.fatal, (message: string) => console.error(bold(red(message))))

const customlogger = new CustomLogger(customLogLevels.error, {
  handlers: [customConsoleHandler]
})

customlogger.trace("trace")
customlogger.debug("debug")
customlogger.info("info")
customlogger.warn("warn")
customlogger.error("error")
customlogger.fatal("fatal")
```

Could we agree on something like this?

**Comment by vwkd** [comment]
@timreichen Your answer doesn't tackle most of the points I gave in my answer. I'll try to keep it short to end the discussion here.

1. While anyone that used all 6 logging levels the framework provided until now will have to write ~30 LOC just to get the base functionality back, you would write exactly 0 LOC more if you just don't use logging levels you don't need.

2. With regards to the default console handler. Now anyone will need to remember, that `Logger()` by default logs to the console, and to turn off logging, you need to pass `{ handlers: [] }` to the `Logger(..)`. This is unintuitive and prone to error. You however would write exactly 1.5 LOC more to get your console handler, which is to import `ConsoleHandler` and initialize it in `Logger({ handlers: [ConsoleHandler] })`. If you also use other handlers, this would probably be what you'd have to do anyways.

3. At last, your preference will **break** anyone that used it until now, while I just advocate for leaving it like it his.

I'll leave it to the repo maintainers to decide, if a purely stylistic alignment to `console` is worth throwing away all of that.

**Pull request closed without merging**
